google_guest_agent/agentcrypto/mtls_mds_windows.go

```diff
@@ -0,0 +1,201 @@
+	// my is predefined personal cert store.
+	my = "MY"
+	// root is predefined cert store for root trusted CA certs.
+	root = "ROOT"
+)
+
+var (
+	prevCtx *windows.CertContext
+)
+
+// writeRootCACert writes Root CA cert from UEFI variable to output file.
+func (j *CredsJob) writeRootCACert(cacert []byte, outputFile string) error {
+	if err := writeToFile(cacert, outputFile); err != nil {
+		return err
+	}
+
+	x509Cert, err := parseCertificate(cacert)
+	if err != nil {
+		return fmt.Errorf("failed to parse root CA cert: %w", err)
+	}
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certcreatecertificatecontext
+	certContext, err := windows.CertCreateCertificateContext(
+		windows.X509_ASN_ENCODING|windows.PKCS_7_ASN_ENCODING,
+		&x509Cert.Raw[0],
+		uint32(len(x509Cert.Raw)))
+	if err != nil {
+		return fmt.Errorf("CertCreateCertificateContext returned: %v", err)
+	}
+	// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certfreecertificatecontext
+	defer windows.CertFreeCertificateContext(certContext)
+
+	// Adds certificate to Root Trusted certificates.
+	if err := addCtxToLocalSystemStore(root, certContext, uint32(windows.CERT_STORE_ADD_REPLACE_EXISTING)); err != nil {
+		return fmt.Errorf("failed to store root cert ctx in store: %w", err)
+	}
+
+	return nil
+}
+
+// writeClientCredentials stores client credentials (certificate and private key).
+func (j *CredsJob) writeClientCredentials(creds []byte, outputFile string) error {
+	if err := writeToFile(creds, outputFile); err != nil {
+		return fmt.Errorf("failed to write client key: %w", err)
+	}
+
+	pfx, err := generatePFX(creds)
+	if err != nil {
+		return fmt.Errorf("failed to generate PFX data from client credentials: %w", err)
+	}
+
+	p := filepath.Join(filepath.Dir(outputFile), pfxFile)
+	if err := writeToFile(pfx, p); err != nil {
+		return fmt.Errorf("failed to write PFX file: %w", err)
+	}
+
+	blob := windows.CryptDataBlob{
+		Size: uint32(len(pfx)),
+		Data: &pfx[0],
+	}
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-pfximportcertstore
+	handle, err := windows.PFXImportCertStore(&blob, syscall.StringToUTF16Ptr(""), uint32(cryptExportable))
+	if err != nil {
+		return fmt.Errorf("failed to import PFX in cert store: %w", err)
+	}
+	defer windows.CertCloseStore(handle, 0)
+
+	var crtCtx *windows.CertContext
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certenumcertificatesinstore
+	crtCtx, err = windows.CertEnumCertificatesInStore(handle, crtCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get cert context for PFX from store: %w", err)
+	}
+	defer windows.CertFreeCertificateContext(crtCtx)
+
+	// Add certificate to personal store.
+	if err := addCtxToLocalSystemStore(my, crtCtx, uint32(windows.CERT_STORE_ADD_NEWER)); err != nil {
+		return fmt.Errorf("failed to store pfx cert context: %w", err)
+	}
+
+	if err := deleteCert(prevCtx, my); err != nil {
```

Comment:
What if we restart the agent we don't know "the previous context"?

Reply:
Added comments to summarize the behavior.